### ISSUE TITLE
NoReturn as generic argument fails to unify with arbitrary type variable in class methods

### ISSUE THREAD

**Issue by schuelermine** [microsoft/pyright#4309]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
`NoReturn` or an arbitrary unused type variable in classes inheriting from generics fails to unify with arbitrary type variables in methods

**To Reproduce**
Try to typecheck the code given below

**Expected behavior**
The code is accepted as correct, or the code is accepted if covariant is changed to contravariant or NoReturn is changed to object. Neither case happens. If both are changed, this code is accepted, though.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from __future__ import annotations
from typing import Generic, NoReturn, TypeVar
T = TypeVar("T", covariant=True)
U = TypeVar("U")
class Foo(Generic[T]):
    pass
def mkBar() -> Bar:
    return Bar()    
def mkFoo() -> Foo[U]:
    return mkBar()
class Bar(Foo[NoReturn]):
    pass
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
VS Code extension, version 1.1.282

**Comment by erictraut** [comment]
Pyright is doing the right thing here. This isn't a bug.

Your code is making improper use of a type variable `U`. This type variable is scoped to the method `mkFoo`, but there's no way for `mkFoo` to return a type of `Foo[U]`. This has nothing to do with the fact that you are using `NoReturn` as a type argument for `Foo` in the `Bar` class definition. The same problem would occur if you replaced `NoReturn` with any other type such as `int` or `str`.

It's highly unusual to use `NoReturn` as a type argument, so I suspect that you probably don't understand what `NoReturn` means. If you provide more details about what you're trying to do here, I can offer more advice.

**Comment by schuelermine** [comment]
As I understand it, `NoReturn` is the bottom type, the type that is a subtype of all other types. With this interpretation, I would expect that a `NoReturn` value can be assigned to any type. Pyright does this in other contexts:
```py
x: int = cast(NoReturn, 12)
```
is checked fine.

**Comment by schuelermine** [comment]
My issue is in the fact that Pyright reports “Type "NoReturn" cannot be assigned to type"U@mkFoo"”, while it obviously can, since `x: int = cast(NoReturn, 12)` works.

**Comment by schuelermine** [comment]
Oh, wait. I realize my mistake. Sorry for bothering you.

**Comment by schuelermine** [comment]
The issue is further illustrated by the fact that this code works:
```py
x: Foo[int]
x = mkBar()
```

**Comment by schuelermine** [comment]
I also understand that the canonical name for the bottom type is `Never`. However, the docs specify that `NoReturn` and `Never` should act identically.

**Comment by erictraut** [comment]
The problem is not that `mkFoo` contains a bare type parameter. The problem is that `Bar` (which is a subtype of `Foo[NoReturn]` is not compatible with type `Foo[U]`. Only `Foo[U]` is compatible with `Foo[U]` in this context.

`Foo[U]` doesn't make sense as a return type of `mkFoo`. Either `Foo[Any]` or `Foo[NoReturn]` might make sense, depending on your intent.

**Comment by schuelermine** [comment]
Assume I take a parameter that decides whether `mkBar` is called, or `Foo` is called. In this case, how would I represent the return type of `mkFoo`?

**Comment by erictraut** [comment]
I don't understand. Can you provide more context? Perhaps some code?

**Comment by schuelermine** [comment]
```py
from __future__ import annotations
from typing import Generic, NoReturn, TypeVar
T = TypeVar("T", covariant=True)
U = TypeVar("U")
class Foo(Generic[T]):
    pass
class Bar(Foo[NoReturn]):
    pass
def mkFoo(x: bool):
    if x:
        return Foo()
    else:
        return Bar()
```
What should I put as the return type annotation on mkFoo?

**Comment by erictraut** [comment]
The type of expression `Foo()` is `Foo[Any]` (or technically `Foo[Unknown]`). The type of expression `Bar()` is `Bar` or `Foo[NoReturn]`. You could annotate the return type as `Foo[Any] | Bar` or `Foo[Any] | Foo[NoReturn]` or simply `Foo[Any]`.

**Comment by schuelermine** [comment]
OK. But if `Foo[NoReturn]` is assignable to `Foo[int]`, in plain script body, why isn't `Foo[NoReturn]` assignable to `Foo[U]`?

**Comment by schuelermine** [comment]
Also, I would expect `Foo[U]` to serve as an alias for all possible types below `Foo`, including `Foo[NoReturn]` (`U` = `NoReturn`)

**Comment by erictraut** [comment]
That's not how type variables work within a function. When a type variable is scoped to a function, all uses of that type variable within that function must use that type variable consistently. If a function signature says that it must return a `Foo[U]`, then the function body must return a `Foo[U]`. It cannot return a `Foo[int]` or `Foo[NoReturn]`.

**Comment by schuelermine** [comment]
But `NoReturn` is a subtype of any type that `U` could be!

**Comment by erictraut** [comment]
You are correct that `NoReturn` is a subtype of any type that `U` can be, but that fact is irrelevant here.

In your earlier code sample, you defined `mkFoo` as `def mkFoo() -> Foo[U]`. This function signature is nonsensical and useless because there's no way for `U` to be solved when it is invoked by a caller. The type variable `U` needs to appear somewhere within an input parameter type, otherwise it cannot be solved.

You corrected this bug in your later code sample by changing `mkFoo` to `def mkFoo(x: U) -> Foo[U]`. Now, the signature makes sense and `U` can be solved when `mkFoo` is called. The type checker must now verify that for any value passed to parameter `x` of type `U`, all code paths return a value that is compatible with `Foo[U]`. If you pass an argument of type `int` to parameter `x`, then the return type must be `Foo[int]`; if you pass an argument of type `float`, then the return type must be `Foo[float]`. This must hold for any type `U`. However, your code doesn't guarantee this. It returns a value of type `Bar`, which is a subtype of `Foo[NoReturn]`, regardless of what the user passed for parameter `x`. That's a type violation because `Foo[NoReturn]` is not consistent with `Foo[U]`.

**Comment by schuelermine** [comment]
So it mustn’t just return a subtype of `Foo[U]`, but an exact match for `Foo[U]`?

**Comment by erictraut** [comment]
Apologies for misdiagnosing this issue. After thinking about it further, I agree with your assessment.

This will be addressed in the next release.

**Comment by schuelermine** [comment]
Thank you

**Comment by erictraut** [comment]
This is included in pyright 1.1.284, which I just published. It will also be included in a future release of pylance.

**Comment by schuelermine** [comment]
Thank you
